exercises/concept/gotta-snatch-em-all/src/test/java/GottaSnatchEmAllTest.java

```diff
@@ -0,0 +1,197 @@
+        GottaSnatchEmAll.addCard("Veevee", collection);
+        assertThat(collection).isEqualTo(expected);
+    }
+
+    @Test
+    @Tag("task:2")
+    @DisplayName("addCard adds the card to the collection when it is a new card")
+    void testAddCardShouldNotAddExistingCardToCollection() {
+        Set<String> collection = new HashSet<>(Set.of("Veevee"));
+        Set<String> expected = Set.of("Veevee");
+        GottaSnatchEmAll.addCard("Veevee", collection);
+        assertThat(collection).isEqualTo(expected);
+    }
+
+    @Test
+    @Tag("task:3")
+    @DisplayName("canTrade returns false when both collections are empty")
+    void testCanTradeBothCollectionsEmpty() {
+        Set<String> myCollection = new HashSet<>();
+        Set<String> theirCollection = new HashSet<>();
+        assertThat(GottaSnatchEmAll.canTrade(myCollection, theirCollection)).isFalse();
+    }
+
+    @Test
+    @Tag("task:3")
+    @DisplayName("canTrade returns false when my collections is empty")
+    void testCanTradeMyCollectionsEmpty() {
+        Set<String> myCollection = new HashSet<>();
+        Set<String> theirCollection = new HashSet<>(Set.of("Bleakachu"));
+        assertThat(GottaSnatchEmAll.canTrade(myCollection, theirCollection)).isFalse();
+    }
+
+    @Test
+    @Tag("task:3")
+    @DisplayName("canTrade returns false when their collections is empty")
+    void testCanTradeTheirCollectionsEmpty() {
+        Set<String> myCollection = new HashSet<>(Set.of("Bleakachu"));
+        Set<String> theirCollection = new HashSet<>();
+        assertThat(GottaSnatchEmAll.canTrade(myCollection, theirCollection)).isFalse();
+    }
+
+    @Test
+    @Tag("task:3")
+    @DisplayName("canTrade returns false when both collections have the same cards")
+    void testCanTradeBothCollectionsHaveSameCards() {
+        Set<String> myCollection = new HashSet<>(Set.of("Gyros", "Garilord"));
+        Set<String> theirCollection = new HashSet<>(Set.of("Garilord", "Gyros"));
+        assertThat(GottaSnatchEmAll.canTrade(myCollection, theirCollection)).isFalse();
+    }
+
+    @Test
+    @Tag("task:3")
+    @DisplayName("canTrade returns true when both collections have unique cards")
+    void testCanTradeBothCollectionsHaveUniqueCards() {
+        Set<String> myCollection = new HashSet<>(Set.of("Gyros"));
+        Set<String> theirCollection = new HashSet<>(Set.of("Garilord"));
+        assertThat(GottaSnatchEmAll.canTrade(myCollection, theirCollection)).isTrue();
+    }
+
+    @Test
+    @Tag("task:3")
+    @DisplayName("canTrade returns true when both collections have at least one card the other doesn't have")
+    void testCanTradeBothCollectionsMixedCards() {
+        Set<String> myCollection = new HashSet<>(Set.of("Gyros", "Garilord", "Bleakachu"));
+        Set<String> theirCollection = new HashSet<>(Set.of("Garilord", "Veevee", "Gyros"));
+        assertThat(GottaSnatchEmAll.canTrade(myCollection, theirCollection)).isTrue();
+    }
+
```

Comment:
Suggest adding a test case for when `myCollection` is just a subset, but `theirCollection` has at least one unique card (and, possibly, when `theirCollection` is a subset of `myCollection`). For example:

```java
        Set<String> myCollection = new HashSet<>(Set.of("Gyros", "Garilord"));
        Set<String> theirCollection = new HashSet<>(Set.of("Garilord", "Veevee", "Gyros"));
```

The instructions (_You cannot trade a card you don't have, and you shouldn't trade a card for one that you already have_ ) seems to imply this should return `true`, but the reverse should return `false`. The test would help to check the sets are checked correctly (for example, in case the student gets `myCollection` and `theirCollection` arguments mixed up).

Reply:
That is a great point, thanks for the suggestion! I believe my reference solution would also trip over these test cases, but I think they make a lot of sense given the instructions. I didn't consider the fact that a trade can be worth it for "me" but not for "them" or vice versa.

Reply:
Added both suggested test cases and updated the reference solution to better match the instructions. And listed you as a contributor, thanks again! 👍